src/framework/MOM_transform_FMS.F90

```diff
@@ -0,0 +1,399 @@
+interface rotated_write_field
+  module procedure rotated_write_field_real_0d
+  module procedure rotated_write_field_real_1d
+  module procedure rotated_write_field_real_2d
+  module procedure rotated_write_field_real_3d
+  module procedure rotated_write_field_real_4d
+end interface rotated_write_field
+
+!> Read a field based on model time, and rotate to the model domain
+interface rotated_time_interp_external
+  module procedure rotated_time_interp_external_0d
+  module procedure rotated_time_interp_external_2d
+  module procedure rotated_time_interp_external_3d
+end interface rotated_time_interp_external
+
+contains
+
+! NOTE: No transformations are applied to the 0d and 1d field implementations,
+!   but are provided to maintain compatibility with the FMS interfaces.
+
+
+!> Compute the FMS (mpp) checksum of a scalar.
+!! This function is provided to support the full FMS mpp_chksum interface.
+function rotated_mpp_chksum_real_0d(field, pelist, mask_val, turns) &
+    result(chksum)
+  real, intent(in) :: field                   !> Input scalar
+  integer, optional, intent(in) :: pelist(:)  !> PE list of ranks to checksum
+  real, optional, intent(in) :: mask_val      !> FMS mask value
+  integer, optional, intent(in) :: turns      !> Number of quarter turns
+  integer :: chksum                           !> FMS checksum of scalar
+
+  if (present(turns)) &
+    call MOM_error(FATAL, "Rotation not supported for 0d fields.")
+
+  chksum = mpp_chksum(field, pelist=pelist, mask_val=mask_val)
+end function rotated_mpp_chksum_real_0d
+
+
+!> Compute the FMS (mpp) checksum of a 1d field.
+!! This function is provided to support the full FMS mpp_chksum interface.
+function rotated_mpp_chksum_real_1d(field, pelist, mask_val, turns) &
+    result(chksum)
+  real, intent(in) :: field(:)                !> Input field
+  integer, optional, intent(in) :: pelist(:)  !> PE list of ranks to checksum
+  real, optional, intent(in) :: mask_val      !> FMS mask value
+  integer, optional, intent(in) :: turns      !> Number of quarter-turns
+  integer :: chksum                           !> FMS checksum of field
+
+  if (present(turns)) &
+    call MOM_error(FATAL, "Rotation not supported for 1d fields.")
+
+  chksum = mpp_chksum(field, pelist=pelist, mask_val=mask_val)
+end function rotated_mpp_chksum_real_1d
+
+
+!> Compute the FMS (mpp) checksum of a rotated 2d field.
+function rotated_mpp_chksum_real_2d(field, pelist, mask_val, turns) &
+    result(chksum)
+  real, intent(in) :: field(:,:)              !> Unrotated input field
+  integer, optional, intent(in) :: pelist(:)  !> PE list of ranks to checksum
+  real, optional, intent(in) :: mask_val      !> FMS mask value
+  integer, optional, intent(in) :: turns      !> Number of quarter-turns
+  integer :: chksum                           !> FMS checksum of field
+
+  real, allocatable :: field_rot(:,:)
+  integer :: qturns
+
+  qturns = 0
+  if (present(turns)) &
+    qturns = turns
```

Comment:
Perhaps this would be even more elegant and robust to user silliness if this line were to read `qturns = modulo(turns, 4)`.

Reply:
I can add it, but I will say that values outside 0-3 are handled by `rotate_array` and `allocate_rotated_array`, so it's probably not necessary in this case.

In fact, I think I have too many `modulo()` calls in the higher level functions, and would ideally only like to handle the modulo at the lowest level.

But happy to add it if you think it will help.

Reply:
It probably doesn't matter too much either way since this is not performance critical code, but since there is an extra allocate and a set of copies, the modulo is probably a cheaper alternative and it would avoid unnecessarily bloating the memory of the model (which might actually matter).  However, if there is already a modulo on the number of turns in the calling code above this, then it does not help.  I would defer to your judgement, @marshallward, on the right approach here.

Reply:
Yes, you are right.. I missed that `qturns = 4` will do the allocate.  Yeah, in that case I will apply the change.